src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -1,14 +1,26 @@
 package org.broadinstitute.dsde.workbench.leonardo.service
 
+import akka.http.scaladsl.model.StatusCodes
 import com.typesafe.scalalogging.LazyLogging
 import org.broadinstitute.dsde.workbench.leonardo.dao.DataprocDAO
-import org.broadinstitute.dsde.workbench.leonardo.db.DbReference
-import org.broadinstitute.dsde.workbench.leonardo.model.{Cluster, ClusterRequest, ClusterResponse}
+import org.broadinstitute.dsde.workbench.leonardo.db.{DataAccess, DbReference}
+import org.broadinstitute.dsde.workbench.leonardo.model.ModelTypes.GoogleProject
+import org.broadinstitute.dsde.workbench.leonardo.model.{Cluster, ClusterRequest, ClusterResponse, LeoException}
+import slick.dbio.DBIO
 
 import scala.concurrent.{ExecutionContext, Future}
 
+case object ClusterNotFoundException extends LeoException
+
 class LeonardoService(gdDAO: DataprocDAO, dbRef: DbReference)(implicit val executionContext: ExecutionContext) extends LazyLogging {
 
+  def withCluster[T](googleProject: GoogleProject, clusterName: String, dataAccess: DataAccess)(op: Cluster => DBIO[T]): DBIO[T] = {
+    dataAccess.clusterQuery.getByName(googleProject, clusterName) flatMap {
+      case None => throw ClusterNotFoundException.toErrorReport(StatusCodes.NotFound)
```

Comment:
consider this a proposal that we start using more descriptive exception classes. insta-wrapping the exception into a `WorkbenchExceptionWithErrorReport` (so the akka-http route exception handler knows what to do with it) seems like an _okay_ mechanism but I'm not super fond of it.

Another option would be to add an optional status code to the `LeoException` base class (default 500), and let exception subclasses define their own (e.g. 404 for `ClusterNotFoundException`). We could then catch `LeoException` in the route exception handler, unpack the status code, and _possibly_ build an `ErrorReport` then instead.

Thoughts?